src/sinks/s3_common/sink.rs

```diff
@@ -69,232 +42,79 @@ where
             service,
             request_builder,
             partitioner,
+            encoding,
+            compression,
             batch_size_bytes,
             batch_size_events,
             batch_timeout,
         }
     }
 }
 
-#[async_trait]
-impl<S, R> StreamSink for S3Sink<S, R>
+impl<Svc, RB, E> S3Sink<Svc, RB, E>
 where
-    S: Service<S3Request> + Send + 'static,
-    S::Future: Send + 'static,
-    S::Response: AsRef<EventStatus> + Send + 'static,
-    S::Error: Debug + Into<crate::Error> + Send,
-    R: S3RequestBuilder + Send,
+    Svc: Service<RB::Request> + Send + 'static,
+    Svc::Future: Send + 'static,
+    Svc::Response: AsRef<EventStatus> + Send + 'static,
+    Svc::Error: fmt::Debug + Into<crate::Error> + Send,
+    RB: RequestBuilder<(String, Vec<Event>)> + Send + Sync + 'static,
+    RB::Events: IntoIterator<Item = Event>,
+    RB::Payload: From<Vec<u8>>,
+    RB::Request: Ackable + Finalizable + Send,
+    E: Encoder + Send + Sync + 'static,
 {
-    async fn run(mut self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
         // All sinks do the same fundamental job: take in events, and ship them
         // out. Empirical testing shows that our number one priority for high
         // throughput needs to be servicing I/O as soon as we possibly can.  In
         // order to do that, we'll spin up a separate task that deals
         // exclusively with I/O, while we deal with everything else here:
         // batching, ordering, and so on.
-        let (io_tx, io_rx) = channel(64);
-        let io_barrier = Arc::new(Barrier::new(2));
-
-        let io = run_io(io_rx, Arc::clone(&io_barrier), self.service, self.acker).in_current_span();
-        let _ = tokio::spawn(io);
-
-        let batcher = Batcher::new(
-            input,
-            self.partitioner,
-            self.batch_timeout,
-            self.batch_size_events,
-            self.batch_size_bytes,
-        );
-        pin!(batcher);
-
-        while let Some((key, batch)) = batcher.next().await {
-            match key {
-                Some(key) => {
-                    // We could push this down to the I/O task if we wanted to.
-                    let request = self.request_builder.build_request(key, batch);
-                    if io_tx.send(request).await.is_err() {
-                        error!(
-                            "Sink I/O channel should not be closed before sink itself is closed."
-                        );
-                        return Err(());
+        let request_builder_rate_limit = NonZeroUsize::new(50);
```

Comment:
small nitpick: This is a concurrency limit, not a rate limit. I'd suggest renaming

Reply:
Good call. Addressed in 05bc5f31a.